core/src/main/scala/org/apache/spark/h2o/converters/ReadConverterContext.scala

```diff
@@ -52,23 +54,43 @@ trait ReadConverterContext {
   def getDouble(columnNum: Int): Option[Double]
   def getString(columnNum: Int): Option[String]
 
-  private type Getter = Int => Option[Any]
-
-  val get: DataType => Getter = Map[DataType, Getter](
-    ByteType -> getByte,
-    ShortType -> getShort,
-    IntegerType -> getInt,
-    LongType -> getLong,
-    FloatType -> getFloat,
-    DoubleType -> getDouble,
-    BooleanType -> getBoolean,
-    StringType -> getUTF8String,
-    TimestampType -> getTimestamp
-  ) withDefault (t => throw new scala.IllegalArgumentException(s"Type $t not supported for conversion from H2OFrame to Spark's Dataframe"))
-
-  def getNullable(dataType: DataType): (Int => Any) = (i: Int) => get(dataType)(i) orNull
-
   def numRows: Int
   def increaseRowIdx() = rowIdx += 1
+
   def hasNext = rowIdx < numRows
+
+  case class OptionReader(name: Any, apply: Int => Option[Any]) {
+  }
+
+//  type OptionReader = Int => Option[Any]
+
+  import SupportedTypes._
+
+  val ReaderPerType = Map[SupportedType, Int => Option[Any]](
+    Byte -> getByte,
+    Short -> getShort,
+    Integer -> getInt,
+    Long -> getLong,
+    Float -> getFloat,
+    Double -> getDouble,
+    Boolean -> getBoolean,
+    String -> getUTF8String,
+    UTF8 -> getUTF8String,
+    Timestamp -> getTimestamp
+  ) withDefault (t => throw new scala.IllegalArgumentException(s"Type $t not supported for conversion from H2OFrame to Spark's Dataframe"))
+
+  def readerFor(dt: DataType): OptionReader = OptionReader(bySparkType(dt), ReaderPerType(bySparkType(dt)))
+
+  case class Reader(name: Any, apply: Int => Any)
+
+  val readerMapByName: Map[NameOfType, Reader]
+
+  /**
+    * For a given array of source column indexes and required data types,
+    * produces an array of value providers.
+    *
+    * @param columnIndexesWithTypes lists which columns we need, and what are the required types
+    * @return an array of value providers. Each provider gives the current column value
+    */
+  def columnValueProviders(columnIndexesWithTypes: Array[(Int, DataType)]): Array[() => Option[Any]]
 }
```

Comment:
`ReadConverterContext` and also `InternalReadConvertorContext` were design in a first way not to depend on whether we convert data to `H2ODataFrame` or `H2ORDD`, which is not not true now because they contain specialised methods to handle Spark's `DataType`s which is only case for conversion from `H2OFrame` to Spark `DataFrame`, not for conversion to `RDD`. 

I would still like to keep them independent on this conversion, right now we split some implementations for H2ORDD and H2ODataframe in multiple classes. You could create another layer on top of ReadConverterContext which would be either for H2ORDD or H2ODataframe